README.md
#strung.lua

Version `1.0.0-rc1`.

a rewrite of the Lua string pattern matching functions in Lua + FFI, for LuaJIT.

`strung.find`, `strung.match`, `strung.gmatch` and `strung.gsub` are implmented according to the Lua manual, and pass the full official test suite.

For the null byte in patterns, `strung` suports both `"%z"`, like Lua 5.1 and `"\0"`, like Lua 5.2. You can capture up to 200 values, if that's your thing.

Like Lua 5.1, and unlike Lua 5.2 and LuaJIT 2.0+, `strung` character classes (`%l`, `%u`, etc.) are locale sensitive.

----

**Contents:** [Usage](#usage) — [Performance](#performance) — [Locales](#locales) — [Undefined Behavior](#undefined-behavior) — [TODO](#todo) — [License](#license) — [Notes](#notes)

----

## Usage

You can use `strung.lua` standalone, the dependencies are optional, and only useful for development.

```Lua
local strung = require"strung"

strung.find("foobar", "b(%l*)") --> 4, 6, "ar"
```

... etc. for `match` and `gmatch`

Alternativley, you can `.install()` the library, and have it replace the original functions completely.

```Lua
strung.install()

print(string.find == strung.find) --> true

S = "foo"

S:match"[^f]*" --> "oo", using `strung.match` rather than `string.match`
```

You can replace the methods selectively by passing method names as string to `install()`.

```Lua
strung.inatall("match", "gmatch")
```

## Performance

### tl;dr

Provided you reuse patterns often:

* For LuaJIT 2.1: `strung` is faster except for `find` with plain strings (which is much slower) and `gsub`.
* For LuaJIT 2.0: `strung` may or may not be competitive, depending on the phase of the moon, except for `gsub`, which is slower. The exact performace is not predictable.

### details

With LuaJIT 2.1, `strung.find`, `.match` and `.gmatch` are consistently faster in my micro benchmarks (except `find()`ing plain strings, see below). `strung.gsub` takes about twice as much time.

With LuaJIT 2.0, depending on the kind of pattern, and on some luck regarding the JIT compiler heuristics [0], matching can be up to three times faster than the original. In other circumstances, for the same pattern, it can be up to three times slower. It is often on par, except for `strung.gsub` which is much slower.

As of 2014-05-09, LuaJIT 2.1 is able to compile calls to `string.find` when the string is not a pattern or when the [`plain` argument](http://www.lua.org/manual/5.1/manual.html#pdf-string.find) is set to `true`. For that use case, the native method is 500-1000 times faster than `strung`, at least for short strings. LuaJIT 2.0 doesn't compile `string.find`, and in that case, `strung` is competitive.

The rest of the string matching functions use the Lua API, and, as such, cause the compiler to abbort if they are in the way. Their `strung` counterpart can be compiled, and included in traces if they are in a hot path.

You have'll to benchmark your peculiar use case to determine if `strung` improves the global performance of your program.

*/!\*: `strung` translates patterns to Lua functions, and caches the result. As a consequence, if you generate a lot of patterns dynamically, and seldom use them, `strung` will be much, much slower than the original. On the other hand, once a pattern has been compiled, matching only depends on the target string, whereas the reference functions have to dispatch on both the pattern and the target string. This allows LuaJIT to compile the matchers optimally.

Plain string search are not compiled

## Locales

`strung` compiles character classes (`"%u"`) and character sets (`"[a-z]"`), and caches the result for each pattern the first time they match. These sets don't update if you change the locale after the fact. You must call `strung.reset()` in order to clear the caches for the locale change to take effect. `strung.setlocale()` is a drop in replacement for `os.setlocale()` that resets `strung`. If you `.install()` the library, `os.setlocale()` will be replaced automatically.

## Undefined behavior

### Bad patterns

`strung` validates the patterns before attempting a match, whereas Lua validates them on the go.

For example:

```Lua
string.find("ab", "^b(") --> nil, the parenthese is never reached.
string.find("ba", "^b(") --> error: unfinished capture
```

`strung` will reject the pattern in both cases.

### Invalid ranges

The interaction between character classes (`%d`) and character ranges (`a-z`) inside character sets (e.g. `[%d-z]`) is documented as undefined in the Lua manual, and `strung` may handle them differently.

Specifically with `strung`:

* When placed before the dash,
  * if `%x` is a character class (e.g. `%l` for lower case letters) `[%l-k]` is a character set containing digits (`%d`), `-` and `k`.
  * If %x is not a character class, the `%x` works as an escape sequence , and thus `[%%-x]` is the character range between `%` and `x`. `[%0-\127]` will match all ASCII characters.
* When the `%` occurs at the end of a character class, it is treated as itself. `[x-%d]` contains the character range between `x` and `%`, and the letter `d`. `[x--]` and `[x-]]` are accepted as ranges ending in `-` and `]`, respectively.

## TODO

* `strung.gsub`
* Boyer Moore for the simple string search?

## License

MIT

## Notes

[0]: In the current benchmark suite, the order of the benchmarks influences the results. for example at some point, testing `gmatch("abcdabcdabcd", "((a)(b)c)()(d)")` alone, strung was taking 1.1 times the time of string.gmatch.

```
-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_
Test:   gmatch  abcdabcdabcd    ((a)(b)c)()(d)
strung/string:  1.1065842049995
```

If you benchmarked `string.find` before `gmatch`, with the same pattern, the result was completely different.

```
-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_
Test:   find    abcdabcdabcd    ((a)(b)c)()(d)
strung/string:  3.1869296949683
-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_
Test:   gmatch  abcdabcdabcd    ((a)(b)c)()(d)
strung/string:  0.29895569825517
```

HISTORY.md
#Change log:

## v1.0

* initial release
* `strung` is now functionally complete, in that it reproduces the API and all features of the Lua string matching library.


tests.lua
local strung = require"strung"
local s_gsub = string.gsub
-- [[]] local v = require"jit.v"
-- [[]] v.on("-")

--- First, the test and benchmark infrastructure.
--- `try` and `gmtry` run both the original and strung version of the functions, 
--- and compare either their ouput, or their speed if "bench" is passed as a 
--- parameter to the script.

local ttstr = require"util".val_to_string
local BASE = 10000
local iter, try, gmtry
if arg[1] == "bench" then 
    function try(f, a, s, d, g, h)
        -- jit.off() jit.on()
        tstring, tStrung = 0, 0
        local ri, Ros
        -- print(("-_"):rep(30))
        -- print("Test: ", )
        local tic = os.clock()
        for i = 1, II do
            ri = {string[f](a, s, d, g, h)}
        end
        tstring = os.clock() - tic
        local tic = os.clock()
        for i = 1, II do
            Ro = {strung[f](a, s, d, g, h)}
        end
        tStrung = os.clock() - tic
        a = s_gsub(a, "[%z\1-\31\127-\255\\\"]", ""):sub(1,60)
        args = '"' .. table.concat({f, a, s, d, g and "true" or nil, h}, '","') ..'"'
        print(table.concat({tStrung/tstring, tStrung, tstring, args}, ","))
    end
    function gmtry(s, p)
        -- print(("-_"):rep(30))
        -- print("Test: ", "gmatch", s, p)
        local ri, ro = {}, {}
        local tic = os.clock()
        for i = 1, II do
            ro = {}
            for a, b, c, d, e, f in strung.gmatch(s, p) do
                ro[#ro + 1] = {a, b, c, d, e, f}
            end
        end
        local tStrung = os.clock() - tic
        local tic = os.clock()
        for i = 1, II do
            ri = {}
            for a, b, c, d, e, f in string.gmatch(s, p) do
                ri[#ri + 1] = {a, b, c, d, e, f}
            end
        end
        local tstring = os.clock() - tic
        s = s_gsub(s, "[%z\1-\31\127-\255\\\"]", ""):sub(1,15)
        args = '"gmatch","'..s..'","'..p..'"'
        print(table.concat({tStrung/tstring, tStrung, tstring, args},","))
    end
    -- function gmtry()end
    function iter(n) II = BASE * n end
    iter(10)
else
    function try(f, ...)
        local params = {...}
        local ri, Ros
        -- print(("-_"):rep(30))
        -- print("Test: ", f, ...)
        ri = {string[f](...)}
        Ro = {strung[f](...)}
        for i, v in ipairs(params) do params[i] = tostring(v) end
        for i = 1, math.max(#ri, #Ro) do
            strung.assert(ri[i] == Ro[i], params[2], table.concat({ 
                table.concat(params, ", "), 
                "ri:", table.concat(ri, ",  "), 
                " \tRo:", table.concat(Ro, ", ")
            }, " | "))
        end
    end
    function gmtry(s, p)
        local desc = "Test:  gmatch ".. s .." -- ".. p
        local ri, ro = {}, {}
        for a, b, c, d, e, f in strung.gmatch(s, p) do
            ro[#ro + 1] = {a, b, c, d, e, f}
        end
        for a, b, c, d, e, f in string.gmatch(s, p) do
            ri[#ri + 1] = {a, b, c, d, e, f}
        end
        strung.assert(#ro == #ri, p, desc.."\nstring: \n"..ttstr(ri).."\n=/=/=/=/=/=/=/=/\nstrung:\n"..ttstr(ro))
        for i = 1, #ro do
        strung.assert(#ro[i] == #ri[i], p, desc.."\nstring: \n"..ttstr(ri).."\n=/=/=/=/=/=/=/=/\nstrung:\n"..ttstr(ro))
            for j = 1, #ri[i] do
                strung.assert(ri[i][j] == ro[i][j], p, desc.."\nstring: \n"..ttstr(ri).."\n=/=/=/=/=/=/=/=/\nstrung:\n"..ttstr(ro))
            end
        end
    end
    iter = function()end
    -- gmtry = iter
end



--- Character classes and locales ---
local allchars do
    local acc = {}
    for i = 0, 255 do acc[i+1] = string.char(i) end
    allchars = table.concat(acc)
end

try("find", allchars, "a")
try("find", allchars, "%a+")
gmtry(allchars, "%a+")

iter(1)

for _, locale in ipairs{
    -- let this out for now, LJ character classes are not sensitive to os.setlocale()
    -- "fr_FR",
    "C"
} do
    -- print("LOCALE: ", strung.setlocale(locale))
    for c in ("acdlpsuwx"):gmatch"." do
        gmtry(allchars, "%"..c.."+")
        gmtry(allchars, "%"..c:upper().."+")
    end
end
gmtry(allchars, "%z+")

-- --- .install() ---

iter(10)

local _f, _m, _gm, _gs, _ol = string.find, string.match, string.gmatch, string.gsub, os.locale

strung.install()
assert(
    string.find == strung.find
    and string.match == strung.match
    and string.gmatch == strung.gmatch
    --and string.gsub == strung.gsub
    and os.locale == strung.locale
    , "`strung.install()` failed.")
--restore the originals.
string.find, string.match, string.gmatch, string.gsub, os.locale = _f, _m, _gm, _gs, _ol


--- The tests ---

assert(strung.find("\0", "\0"), "'\\0' pattern failed to match.")
assert(strung.find("\0", "[\0]"), "'\\0' pattern failed to match in charset.")


--- %f ---

try("find", "\0", "%z")
try("find", "\0", "[%z]")

try("find", "", "[%z]")


try("find", "a", "a?a")

try("find", "AAAAAA", "%f[%l]a")
try("find", "AAAAAA", "%f[%l]")
try("find", "aAaAb", "%f[%l]a", 2)
try("find", "aAaAb", "%f[%l]a")
try("find", "aAaAb", "%f[%l]a", 4)
try("find", "AaAb", "%f[%l]a")
try("find", "aAb", "%f[%l]b")
try("find", "aAb", "%f[%l]a")


--- negative indices ---

try("find", "fof", "f", -4)
try("find", "fof", "f", -3)
try("find", "fof", "f", -2)
try("find", "fof", "f", -1)

try("find", "fof", "[^o]", -4)
try("find", "fof", "[^o]", -3)
try("find", "fof", "[^o]", -2)
try("find", "fof", "[^o]", -1)

try("match", "fof", "f", -4)
try("match", "fof", "f", -3)
try("match", "fof", "f", -2)
try("match", "fof", "f", -1)

try("match", "fof", "[^o]", -4)
try("match", "fof", "[^o]", -3)
try("match", "fof", "[^o]", -2)
try("match", "fof", "[^o]", -1)

--- gmatch ---

gmtry('abcdabcdabcd', "((a)(b)c)()(d)")
-- try("find", 'abcdabcdabcd', "((a)(b)c)()(d)")
-- try("find", 'abcdabcdabcd', "(a)(b)c(d)")
gmtry('abcdabcdabcd', "(a)(b)c(d)")
gmtry('abcdabcdabcd', "(a)(b)(d)")
gmtry('abcdabcdabcd', "(a)(b)(d)")
gmtry('abcabcabc', "(a)(b)")
gmtry('abcabcabc', "(ab)")

iter(10)
--- bug fix ---

try("match", "faa:foo:", "(.+):(%l+)")
try("match", ":foo:", "(%l*)")
try("match", "faa:foo:", ":%l+")
try("match", "faa:foo:", ":(%l+)")
try("match", "faa:foo:", "(%l+)")
try("match", ":foo:", "(%l+)")
try("match", "foo", "%l+")
try("match", "foo", "foo")

--- anchored patterns ---

try("find", "wwS", "^wS", 2)
try("find", "wwS", "^wS")
try("find", "wwS", "^ww", 2)
try("find", "wwS", "^ww")

--- %b ---

try("find", "a(f()g(h(d d))[[][]]K)", "%b()%b[]", 3)
try("find", "a(f()g(h(d d))[[][]]K)", "%b()%b[]", 2)
try("find", "a(f()g(h(d d))[[][]]K)", "%b()%b[]")
try("find", "a(f()g(h(d d))K)", "%b()")
try("find", "a(f()g(h(d d))K", "%b()")
try("find", "foobarfoo", "(foo)(bar)%2%1")
try("find", "foobarbarfoo", "(foo)(bar)%2%1")
try("find", "foobarbar", "(foo)(bar)%2")
try("find", "foobarfoo", "(foo)(bar)%2")
try("find", "foobarfoo", "(foo)(bar)%1")
try("find", "foobarfoo", "(foo)bar%1")

--- Captures ---

try("find", "wwS", "((w*)S)")
try("find", "wwwwS", "((w*)%u)")
try("find", "wwS", "((%l)%u)")
try("find", "SSw", "((%u)%l)")
try("find", "wwS", "((%l*)%u)")
try("find", "wwS", "((%l-)%u)")

try("find", "wwS", "((w*)%u)")
try("find", "wwS", "((ww)%u)")
try("find", "wwS", "((%l*)S)")
try("find", "wwS", "((%l*))")


try("find", "wwSS", "()(%u+)()")

--- Character sets  ---

try("find", "wwwwwwS", "[^%u]*")
try("find", "wwwwwwS", "[^%u]")
try("find", "wwwwwwS", "(%l*)")
try("find", "wwSS", "(%u+)")

try("find", "wwS", "%l%u")
try("find", "wwS", "()%l%u")

try("find", "wwwwwwS", "[^%U]")
try("find", "wwwwwwS", "[%U]*")
try("find", "wwwwwwS", "[%U]+")

try("find", "wwwwwwS", "%l*")

try("find", "wwwwwwS", "[%U]")
try("find", "wwwwwwS", "[%u]")

try("find", "wwwwwwS", "[%u]*")
try("find", "wwwwwwS", "[^kfdS]*")

try("find", "wwS", "%l*()")
try("find", "wwS", "()%u+")

--- escape sequeces ---

try("find", "w(wSESDFB)SFwe)fwe", "%(.-%)")
try("find", "w(wSESDFB)SFwe)fwe", "%(.*%)")

--- Basic patterns ---

try("find", "wawSESDFB)SFweafwe", "a.-a")
try("find", "wawSESDFBaSFwe)fwe", "a.*a")

try("find", "a", ".")
try("find", "a6ruyfhjgjk9", ".+")



try("find", "wawSESDFBaSFwe)fwe", "a[A-Za-z]*a")

try("find", "qwwSYUGJHDwefwe", "%u+")
try("find", "wwSESDFBSFwefwe", "[A-Z]+")

try("find", "SYUGJHD", "%u+")
try("find", "SESDFBSF", "[A-Z]+")
try("find", "qwwSYUGJHD", "%u+")
try("find", "wwSESDFBSF", "[A-Z]+")

try("find", "S", "%u")
try("find", "S", "[A-Z]")

try("find", "ab", "a?b")
try("find", "b", "a?b")
try("find", "abbab", "a?ba?b$")
try("find", "abbabbab", "a?ba?b$")
try("find", "abbabbab", "a?ba?ba?ba?ba?b$")
try("find", "abbabbaba", "a?ba?ba?ba?ba?b$")

try("find", "aaaabaaaaabbaaaabb$", "a+bb$")
try("find", "aaaabaaaaabbaaaabb", "a*bb$")
try("find", "aaaaaaaabaaabaaaaabb", "a+bb")
try("find", "aaaaaaaabaaabaaaaabb", "a*bb")

try("find", "aaaaaaaabaaabaaaaab", "ba-bb")
try("find", "aaaaaaaabaaabaaaaabb", "ba-bb")

try("find", "aaa", "a+")
try("find", "aaaaaaaaaaaaaaaaaa", "a+")

-- iter(0.0001)
-- try("find", ("aaaaaaaaabaaaaaaaabbaaaaaaaaaaaabbaaaaaaaaaaabbaaaaaaaaaaaabb"):rep(10000), "aaaaaaaaaaaaabbb")

-- s = {}
-- for i = 1, 610000 do
--     s[#s+1] = string.char(math.random(255))
-- end
-- s = table.concat(s)
-- collectgarbage()
-- try("find", s, "aaaaaaaaaaaaabbb")

try("find", "aaaaabaaaaabaaaaaaaaabb", "aabb")
try("find", "aaaaaaaaabbaaaaaaaabbaaaaaaaaaaaabbaaaaaaaaaaabbaaaaaaaaaaaabb", "aaaaaaaaaaaaabbb")

iter(10)

try("find", "baa", "aa")
try("find", "ba", "a")

try("find", "a", "aa")
try("find", "aa", "a")
try("find", "aa", "aa")
try("find", "a", "a")

try("find", "aaaaabaaaaabaaaaaaaaabb", "aabb", nil, true)
try("find", "aaaaaaaaabbaaaaaaaabbaaaaaaaaaaaabbaaaaaaaaaaabbaaaaaaaaaaaabb", "aaaaaaaaaaaaabbb", nil, true)

iter(10)

try("find", "baa", "aa", nil, true)
try("find", "ba", "a", nil, true)

try("find", "a", "aa", nil, true)
try("find", "aa", "a", nil, true)
try("find", "aa", "aa", nil, true)
try("find", "a", "a", nil, true)



if arg[1] ~= "bench" then print "ok" end

tests/gsub.lua
--- gsub ---

local c = require"tests.common"
local try, gmtry, iter, allchars, dumpacc = c.try, c.gmtry, c.iter, c.allchars, c.dumpacc

iter(10)

-- basic tests

try("gsub", "_d_d_", "d", "+")
try("gsub", "_da_da_", "(d)a", "+")

try("gsub", "_d_d_", "d", {})
try("gsub", "_d_d_", "d", {d = 9})
try("gsub", "_d_d_", "d", {d = "9"})

try("gsub", "_d_d_", "d", function()end)
try("gsub", "_do_d_", "(d)(.)", function(a,b) return b,a end)

-- the patterns used by f1() in pm.lua

try('gsub', '123abc123', "abc", "IVX%1IVX")

try("gsub", '(..*) %1', "%%([0-9])", function (s) return "%" .. (s+1) end)
try("gsub", '(..*) %1', "^(^?)", "%1()", 1)
try("gsub", '(..*) %1', "($?)$", "()%1", 1)

-- see issue #7

try("gsub", "os:d:/dropbox/goluwa/.userdata/caps/lua/libraries/extensions/globals.lua", "^(.-:)", "")

if not bench then print"ok" end


pm.lua
local strung = require"strung"

print('testing pattern matching')

function f(s, p)
  local i,e = strung.find(s, p)
  if i then return strung.sub(s, i, e) end
end

function f1(s, p)
  p = strung.gsub(p, "%%([0-9])", function (s) return "%" .. (s+1) end)
  p = strung.gsub(p, "^(^?)", "%1()", 1)
  p = strung.gsub(p, "($?)$", "()%1", 1)
  local t = {strung.match(s, p)}
  return strung.sub(s, t[1], t[#t] - 1)
end

a,b = strung.find('', '')    -- empty patterns are tricky
assert(a == 1 and b == 0);
a,b = strung.find('alo', '')
assert(a == 1 and b == 0)
a,b = strung.find('a\0o a\0o a\0o', 'a', 1)   -- first position
assert(a == 1 and b == 1)
a,b = strung.find('a\0o a\0o a\0o', 'a\0o', 2)   -- starts in the midle
assert(a == 5 and b == 7)
a,b = strung.find('a\0o a\0o a\0o', 'a\0o', 9)   -- starts in the midle
assert(a == 9 and b == 11)
a,b = strung.find('a\0a\0a\0a\0\0ab', '\0ab', 2);  -- finds at the end
assert(a == 9 and b == 11);
a,b = strung.find('a\0a\0a\0a\0\0ab', 'b')    -- last position
assert(a == 11 and b == 11)
assert(strung.find('a\0a\0a\0a\0\0ab', 'b\0') == nil)   -- check ending
assert(strung.find('', '\0') == nil)
assert(strung.find('alo123alo', '12') == 4)
assert(strung.find('alo123alo', '^12') == nil)

-- assert(f('aloALO', '%l*') == 'alo')
-- assert(f('aLo_ALO', '%a*') == 'aLo')

-- assert(f('aaab', 'a*') == 'aaa');
-- assert(f('aaa', '^.*$') == 'aaa');
-- assert(f('aaa', 'b*') == '');
-- assert(f('aaa', 'ab*a') == 'aa')
-- assert(f('aba', 'ab*a') == 'aba')
-- assert(f('aaab', 'a+') == 'aaa')
-- assert(f('aaa', '^.+$') == 'aaa')
-- assert(f('aaa', 'b+') == nil)
-- assert(f('aaa', 'ab+a') == nil)
-- assert(f('aba', 'ab+a') == 'aba')
-- assert(f('a$a', '.$') == 'a')
-- assert(f('a$a', '.%$') == 'a$')
-- assert(f('a$a', '.$.') == 'a$a')
-- assert(f('a$a', '$$') == nil)
-- assert(f('a$b', 'a$') == nil)
-- assert(f('a$a', '$') == '')
-- assert(f('', 'b*') == '')
-- assert(f('aaa', 'bb*') == nil)
-- assert(f('aaab', 'a-') == '')
-- assert(f('aaa', '^.-$') == 'aaa')
-- assert(f('aabaaabaaabaaaba', 'b.*b') == 'baaabaaabaaab')
-- assert(f('aabaaabaaabaaaba', 'b.-b') == 'baaab')
-- assert(f('alo xo', '.o$') == 'xo')
-- assert(f(' \n isto ? assim', '%S%S*') == 'isto')
-- assert(f(' \n isto ? assim', '%S*$') == 'assim')
-- assert(f(' \n isto ? assim', '[a-z]*$') == 'assim')
-- assert(f('um caracter ? extra', '[^%sa-z]') == '?')
-- assert(f('', 'a?') == '')
-- assert(f('?', '??') == '?')
-- assert(f('?bl', '??b?l?') == '?bl')
-- assert(f('  ?bl', '??b?l?') == '')
-- assert(f('aa', '^aa?a?a') == 'aa')
-- assert(f(']]]?b', '[^]]') == '?')
-- assert(f("0alo alo", "%x*") == "0a")
-- assert(f("alo alo", "%C+") == "alo alo")
-- print('+')

-- assert(f1('alo alx 123 b\0o b\0o', '(..*) %1') == "b\0o b\0o")
-- assert(f1('axz123= 4= 4 34', '(.+)=(.*)=%2 %1') == '3= 4= 4 3')
-- assert(f1('=======', '^(=*)=%1$') == '=======')
assert(strung.match('==========', '^([=]*)=%1$') == nil)

local function range (i, j)
  if i <= j then
    return i, range(i+1, j)
  end
end

local abc = string.char(range(0, 255));

assert(string.len(abc) == 256)

-- function strset (p)
--   local res = {s=''}
--   strung.gsub(abc, p, function (c) res.s = res.s .. c end)
--   return res.s
-- end;

-- assert(strung.len(strset('[\200-\210]')) == 11)

-- assert(strset('[a-z]') == "abcdefghijklmnopqrstuvwxyz")
-- assert(strset('[a-z%d]') == strset('[%da-uu-z]'))
-- assert(strset('[a-]') == "-a")
-- assert(strset('[^%W]') == strset('[%w]'))
-- assert(strset('[]%%]') == '%]')
-- assert(strset('[a%-z]') == '-az')
-- assert(strset('[%^%[%-a%]%-b]') == '-[]^ab')
-- assert(strset('%Z') == strset('[\1-\255]'))
-- assert(strset('.') == strset('[\1-\255%z]'))
-- print('+');

assert(strung.match("alo xyzK", "(%w+)K") == "xyz")
assert(strung.match("254 K", "(%d*)K") == "")
--[[]]
-- assert(strung.match("alo ", "(%w*)$") == "")
assert(strung.match("alo ", "(%w+)$") == nil)
assert(strung.find("(?lo)", "%(?") == 1)
local a, b, c, d, e = strung.match("?lo alo", "^(((.).).* (%w*))$")
assert(a == '?lo alo' and b == '?l' and c == '?' and d == 'alo' and e == nil)
a, b, c, d  = strung.match('0123456789', '(.+(.?)())')
assert(a == '0123456789' and b == '' and c == 11 and d == nil)
print('+')

-- assert(strung.gsub('?lo ?lo', '?', 'x') == 'xlo xlo')
-- assert(strung.gsub('alo ?lo  ', ' +$', '') == 'alo ?lo')  -- trim
-- assert(strung.gsub('  alo alo  ', '^%s*(.-)%s*$', '%1') == 'alo alo')  -- double trim
-- assert(strung.gsub('alo  alo  \n 123\n ', '%s+', ' ') == 'alo alo 123 ')
-- t = "ab? d"
-- a, b = strung.gsub(t, '(.)', '%1@')
-- assert('@'..a == strung.gsub(t, '', '@') and b == 5)
-- a, b = strung.gsub('ab?d', '(.)', '%0@', 2)
-- assert(a == 'a@b@?d' and b == 2)
-- assert(strung.gsub('alo alo', '()[al]', '%1') == '12o 56o')
-- assert(strung.gsub("abc=xyz", "(%w*)(%p)(%w+)", "%3%2%1-%0") ==
--               "xyz=abc-abc=xyz")
-- assert(strung.gsub("abc", "%w", "%1%0") == "aabbcc")
-- assert(strung.gsub("abc", "%w+", "%0%1") == "abcabc")
-- assert(strung.gsub('???', '$', '\0??') == '???\0??')
-- assert(strung.gsub('', '^', 'r') == 'r')
-- assert(strung.gsub('', '$', 'r') == 'r')
-- print('+')

-- assert(strung.gsub("um (dois) tres (quatro)", "(%(%w+%))", strung.upper) ==
--             "um (DOIS) tres (QUATRO)")

-- do
--   local function setglobal (n,v) rawset(_G, n, v) end
--   strung.gsub("a=roberto,roberto=a", "(%w+)=(%w%w*)", setglobal)
--   assert(_G.a=="roberto" and _G.roberto=="a")
-- end

-- function f(a,b) return strung.gsub(a,'.',b) end
-- assert(strung.gsub("trocar tudo em |teste|b| ? |beleza|al|", "|([^|]*)|([^|]*)|", f) ==
--             "trocar tudo em bbbbb ? alalalalalal")

local function dostrung (s) return loadstrung(s)() or "" end
-- assert(strung.gsub("alo $a=1$ novamente $return a$", "$([^$]*)%$", dostrung) ==
--             "alo  novamente 1")

-- x = strung.gsub("$x=strung.gsub('alo', '.', strung.upper)$ assim vai para $return x$",
--          "$([^$]*)%$", dostrung)
-- assert(x == ' assim vai para ALO')

-- t = {}
-- s = 'a alo jose  joao'
-- r = strung.gsub(s, '()(%w+)()', function (a,w,b)
--       assert(strung.len(w) == b-a);
--       t[a] = b-a;
--     end)
-- assert(s == r and t[1] == 1 and t[3] == 3 and t[7] == 4 and t[13] == 4)


-- function isbalanced (s)
--   return strung.find(strung.gsub(s, "%b()", ""), "[()]") == nil
-- end

-- assert(isbalanced("(9 ((8))(\0) 7) \0\0 a b ()(c)() a"))
-- assert(not isbalanced("(9 ((8) 7) a b (\0 c) a"))
-- assert(strung.gsub("alo 'oi' alo", "%b''", '"') == 'alo " alo')


local t = {"apple", "orange", "lime"; n=0}
-- assert(strung.gsub("x and x and x", "x", function () t.n=t.n+1; return t[t.n] end)
--         == "apple and orange and lime")

-- t = {n=0}
-- strung.gsub("first second word", "%w%w*", function (w) t.n=t.n+1; t[t.n] = w end)
-- assert(t[1] == "first" and t[2] == "second" and t[3] == "word" and t.n == 3)

-- t = {n=0}
-- assert(strung.gsub("first second word", "%w+",
--          function (w) t.n=t.n+1; t[t.n] = w end, 2) == "first second word")
-- assert(t[1] == "first" and t[2] == "second" and t[3] == nil)

-- assert(not pcall(strung.gsub, "alo", "(.", print))
-- assert(not pcall(strung.gsub, "alo", ".)", print))
-- assert(not pcall(strung.gsub, "alo", "(.", {}))
-- assert(not pcall(strung.gsub, "alo", "(.)", "%2"))
-- assert(not pcall(strung.gsub, "alo", "(%1)", "a"))
-- assert(not pcall(strung.gsub, "alo", "(%0)", "a"))

-- big strungs
local a = string.rep('a', 300000)
assert(strung.find(a, '^a*.?$'))
assert(not strung.find(a, '^a*.?b$'))
assert(strung.find(a, '^a-.?$'))

-- deep nest of gsubs
-- function rev (s)
--   return strung.gsub(s, "(.)(.+)", function (c,s1) return rev(s1)..c end)
-- end

-- local x = strung.rep('012345', 10)
-- assert(rev(rev(x)) == x)


-- -- gsub with tables
-- assert(strung.gsub("alo alo", ".", {}) == "alo alo")
-- assert(strung.gsub("alo alo", "(.)", {a="AA", l=""}) == "AAo AAo")
-- assert(strung.gsub("alo alo", "(.).", {a="AA", l="K"}) == "AAo AAo")
-- assert(strung.gsub("alo alo", "((.)(.?))", {al="AA", o=false}) == "AAo AAo")

-- assert(strung.gsub("alo alo", "().", {2,5,6}) == "256 alo")

-- t = {}; setmetatable(t, {__index = function (t,s) return strung.upper(s) end})
-- assert(strung.gsub("a alo b hi", "%w%w+", t) == "a ALO b HI")


-- tests for gmatch
assert(strung.gfind == strung.gmatch)
local a = 0
--[[]]
-- for i in strung.gmatch('abcde', '()') do 
--     print(i)
--     assert(i == a+1); a=i
-- end
-- assert(a==6)

t = {n=0}
for w in strung.gmatch("first second word", "%w+") do
      t.n=t.n+1; t[t.n] = w
end
assert(t[1] == "first" and t[2] == "second" and t[3] == "word")

t = {3, 6, 9}
for i in strung.gmatch ("xuxx uu ppar r", "()(.)%2") do
  assert(i == table.remove(t, 1))
end
assert(table.getn(t) == 0)

t = {}
for i,j in strung.gmatch("13 14 10 = 11, 15= 16, 22=23", "(%d+)%s*=%s*(%d+)") do
  t[i] = j
end
a = 0
for k,v in pairs(t) do assert(k+1 == v+0); a=a+1 end
assert(a == 3)


-- -- tests for `%f' (`frontiers')

-- assert(strung.gsub("aaa aa a aaa a", "%f[%w]a", "x") == "xaa xa x xaa x")
-- assert(strung.gsub("[[]] [][] [[[[", "%f[[].", "x") == "x[]] x]x] x[[[")
-- assert(strung.gsub("01abc45de3", "%f[%d]", ".") == ".01abc.45de.3")
-- assert(strung.gsub("01abc45 de3x", "%f[%D]%w", ".") == "01.bc45 de3.")
-- assert(strung.gsub("function", "%f[\1-\255]%w", ".") == ".unction")
-- assert(strung.gsub("function", "%f[^\1-\255]", ".") == "function.")

local i, e = strung.find(" alo aalo allo", "%f[%S].-%f[%s].-%f[%S]")
assert(i == 2 and e == 5)
local k = strung.match(" alo aalo allo", "%f[%S](.-%f[%s].-%f[%S])")
assert(k == 'alo ')

--[[]]
-- local a = {1, 5, 9, 14, 17,}
-- for k in strung.gmatch("alo alo th02 is 1hat", "()%f[%w%d]") do
--   print(k, type(k))
--   assert(table.remove(a, 1) == k)
-- end
-- assert(table.getn(a) == 0)


print('OK')
